### ISSUE TITLE
language server crashed

### ISSUE THREAD

**Issue by GabrielHPDS** [redhat-developer/vscode-xml#630]
"The XML language server crashed 5 times in the last 10 minutes. The server will not be restarted."

I've tried everything, including uninstalled and reinstalled the extension and also vscode but the same error continues.

**Comment by angelozerr** [comment]
Could you enable xml trace https://github.com/redhat-developer/vscode-xml/blob/master/docs/Troubleshooting.md#no-support-on-xml-file and share your trace please

**Comment by angelozerr** [comment]
@GabrielHPDS is there any chance to have your traces?

**Comment by angelozerr** [comment]
None answer. I close this issue. @GabrielHPDS if you can reproduce the problem and send us trace and it will be very helfull. Thanks!

**Comment by Torsteinws** [comment]
I have the same exact problem. Here is a copy of my trace:

```
Params: {
    "processId": 20288,
    "clientInfo": {
        "name": "Visual Studio Code",
        "version": "1.64.0"
    },
    "locale": "en-us",
    "rootPath": "c:\\Users\\MyUser\\MyProject",
    "rootUri": "file:///c%3A/Users/MyUser/MyProject",
    "capabilities": {
        "workspace": {
            "applyEdit": true,
            "workspaceEdit": {
                "documentChanges": true,
                "resourceOperations": [
                    "create",
                    "rename",
                    "delete"
                ],
                "failureHandling": "textOnlyTransactional",
                "normalizesLineEndings": true,
                "changeAnnotationSupport": {
                    "groupsOnLabel": true
                }
            },
            "didChangeConfiguration": {
                "dynamicRegistration": true
            },
            "didChangeWatchedFiles": {
                "dynamicRegistration": true
            },
            "symbol": {
                "dynamicRegistration": true,
                "symbolKind": {
                    "valueSet": [
                        1,
                        2,
                        3,
                        4,
                        5,
                        6,
                        7,
                        8,
                        9,
                        10,
                        11,
                        12,
                        13,
                        14,
                        15,
                        16,
                        17,
                        18,
                        19,
                        20,
                        21,
                        22,
                        23,
                        24,
                        25,
                        26
                    ]
                },
                "tagSupport": {
                    "valueSet": [
                        1
                    ]
                }
            },
            "codeLens": {
                "refreshSupport": true
            },
            "executeCommand": {
                "dynamicRegistration": true
            },
            "configuration": true,
            "workspaceFolders": true,
            "semanticTokens": {
                "refreshSupport": true
            },
            "fileOperations": {
                "dynamicRegistration": true,
                "didCreate": true,
                "didRename": true,
                "didDelete": true,
                "willCreate": true,
                "willRename": true,
                "willDelete": true
            }
        },
        "textDocument": {
            "publishDiagnostics": {
                "relatedInformation": true,
                "versionSupport": false,
                "tagSupport": {
                    "valueSet": [
                        1,
                        2
                    ]
                },
                "codeDescriptionSupport": true,
                "dataSupport": true
            },
            "synchronization": {
                "dynamicRegistration": true,
                "willSave": true,
                "willSaveWaitUntil": true,
                "didSave": true
            },
            "completion": {
                "dynamicRegistration": true,
                "contextSupport": true,
                "completionItem": {
                    "snippetSupport": true,
                    "commitCharactersSupport": true,
                    "documentationFormat": [
                        "markdown",
                        "plaintext"
                    ],
                    "deprecatedSupport": true,
                    "preselectSupport": true,
                    "tagSupport": {
                        "valueSet": [
                            1
                        ]
                    },
                    "insertReplaceSupport": true,
                    "resolveSupport": {
                        "properties": [
                            "documentation",
                            "detail",
                            "additionalTextEdits"
                        ]
                    },
                    "insertTextModeSupport": {
                        "valueSet": [
                            1,
                            2
                        ]
                    }
                },
                "completionItemKind": {
                    "valueSet": [
                        1,
                        2,
                        3,
                        4,
                        5,
                        6,
                        7,
                        8,
                        9,
                        10,
                        11,
                        12,
                        13,
                        14,
                        15,
                        16,
                        17,
                        18,
                        19,
                        20,
                        21,
                        22,
                        23,
                        24,
                        25
                    ]
                }
            },
            "hover": {
                "dynamicRegistration": true,
                "contentFormat": [
                    "markdown",
                    "plaintext"
                ]
            },
            "signatureHelp": {
                "dynamicRegistration": true,
                "signatureInformation": {
                    "documentationFormat": [
                        "markdown",
                        "plaintext"
                    ],
                    "parameterInformation": {
                        "labelOffsetSupport": true
                    },
                    "activeParameterSupport": true
                },
                "contextSupport": true
            },
            "definition": {
                "dynamicRegistration": true,
                "linkSupport": true
            },
            "references": {
                "dynamicRegistration": true
            },
            "documentHighlight": {
                "dynamicRegistration": true
            },
            "documentSymbol": {
                "dynamicRegistration": true,
                "symbolKind": {
                    "valueSet": [
                        1,
                        2,
                        3,
                        4,
                        5,
                        6,
                        7,
                        8,
                        9,
                        10,
                        11,
                        12,
                        13,
                        14,
                        15,
                        16,
                        17,
                        18,
                        19,
                        20,
                        21,
                        22,
                        23,
                        24,
                        25,
                        26
                    ]
                },
                "hierarchicalDocumentSymbolSupport": true,
                "tagSupport": {
                    "valueSet": [
                        1
                    ]
                },
                "labelSupport": true
            },
            "codeAction": {
                "dynamicRegistration": true,
                "isPreferredSupport": true,
                "disabledSupport": true,
                "dataSupport": true,
                "resolveSupport": {
                    "properties": [
                        "edit"
                    ]
                },
                "codeActionLiteralSupport": {
                    "codeActionKind": {
                        "valueSet": [
                            "",
                            "quickfix",
                            "refactor",
                            "refactor.extract",
                            "refactor.inline",
                            "refactor.rewrite",
                            "source",
                            "source.organizeImports"
                        ]
                    }
                },
                "honorsChangeAnnotations": false
            },
            "codeLens": {
                "dynamicRegistration": true
            },
            "formatting": {
                "dynamicRegistration": true
            },
            "rangeFormatting": {
                "dynamicRegistration": true
            },
            "onTypeFormatting": {
                "dynamicRegistration": true
            },
            "rename": {
                "dynamicRegistration": true,
                "prepareSupport": true,
                "prepareSupportDefaultBehavior": 1,
                "honorsChangeAnnotations": true
            },
            "documentLink": {
                "dynamicRegistration": true,
                "tooltipSupport": true
            },
            "typeDefinition": {
                "dynamicRegistration": true,
                "linkSupport": true
            },
            "implementation": {
                "dynamicRegistration": true,
                "linkSupport": true
            },
            "colorProvider": {
                "dynamicRegistration": true
            },
            "foldingRange": {
                "dynamicRegistration": true,
                "rangeLimit": 5000,
                "lineFoldingOnly": true
            },
            "declaration": {
                "dynamicRegistration": true,
                "linkSupport": true
            },
            "selectionRange": {
                "dynamicRegistration": true
            },
            "callHierarchy": {
                "dynamicRegistration": true
            },
            "semanticTokens": {
                "dynamicRegistration": true,
                "tokenTypes": [
                    "namespace",
                    "type",
                    "class",
                    "enum",
                    "interface",
                    "struct",
                    "typeParameter",
                    "parameter",
                    "variable",
                    "property",
                    "enumMember",
                    "event",
                    "function",
                    "method",
                    "macro",
                    "keyword",
                    "modifier",
                    "comment",
                    "string",
                    "number",
                    "regexp",
                    "operator"
                ],
                "tokenModifiers": [
                    "declaration",
                    "definition",
                    "readonly",
                    "static",
                    "deprecated",
                    "abstract",
                    "async",
                    "modification",
                    "documentation",
                    "defaultLibrary"
                ],
                "formats": [
                    "relative"
                ],
                "requests": {
                    "range": true,
                    "full": {
                        "delta": true
                    }
                },
                "multilineTokenSupport": false,
                "overlappingTokenSupport": false
            },
            "linkedEditingRange": {
                "dynamicRegistration": true
            }
        },
        "window": {
            "showMessage": {
                "messageActionItem": {
                    "additionalPropertiesSupport": true
                }
            },
            "showDocument": {
                "support": true
            },
            "workDoneProgress": true
        },
        "general": {
            "regularExpressions": {
                "engine": "ECMAScript",
                "version": "ES2020"
            },
            "markdown": {
                "parser": "marked",
                "version": "1.1.0"
            }
        }
    },
    "initializationOptions": {
        "settings": {
            "xml": {
                "java": {
                    "home": "C:\\Applications\\Atlassian\\atlassian-plugin-sdk-8.0.16\\jre"
                },
                "server": {
                    "vmargs": "-Xmx64M",
                    "workDir": "~/.lemminx",
                    "preferBinary": false,
                    "silenceExtensionWarning": false,
                    "binary": {
                        "path": "",
                        "args": "",
                        "trustedHashes": []
                    }
                },
                "trace": {
                    "server": "verbose"
                },
                "logs": {
                    "client": true,
                    "file": "c:\\Users\\MyUser\\AppData\\Roaming\\Code\\User\\workspaceStorage\\46873f98d2b0283b1e5a9d959af73d01\\redhat.vscode-xml\\lemminx.log"
                },
                "downloadExternalResources": {
                    "enabled": true
                },
                "catalogs": [],
                "fileAssociations": [],
                "format": {
                    "enabled": true,
                    "emptyElements": "ignore",
                    "enforceQuoteStyle": "ignore",
                    "joinCDATALines": false,
                    "joinCommentLines": false,
                    "joinContentLines": false,
                    "preserveAttributeLineBreaks": false,
                    "preserveEmptyContent": false,
                    "preservedNewlines": 2,
                    "spaceBeforeEmptyCloseTag": true,
                    "xsiSchemaLocationSplit": "none",
                    "splitAttributes": false,
                    "splitAttributesIndentSize": 2,
                    "closingBracketNewLine": false,
                    "trimFinalNewlines": false,
                    "trimTrailingWhitespace": false,
                    "insertFinalNewline": false
                },
                "completion": {
                    "autoCloseTags": true,
                    "autoCloseRemovesContent": true
                },
                "codeLens": {
                    "enabled": false
                },
                "preferences": {
                    "quoteStyle": "double",
                    "showSchemaDocumentationType": "all"
                },
                "validation": {
                    "enabled": true,
                    "namespaces": {
                        "enabled": "always"
                    },
                    "schema": {
                        "enabled": "always"
                    },
                    "disallowDocTypeDecl": false,
                    "resolveExternalEntities": false,
                    "noGrammar": "hint"
                },
                "symbols": {
                    "enabled": true,
                    "excluded": [],
                    "maxItemsComputed": 5000,
                    "showReferencedGrammars": true,
                    "filters": []
                },
                "extension": {
                    "jars": []
                },
                "useCache": true,
                "telemetry": {
                    "enabled": true
                }
            }
        },
        "extendedClientCapabilities": {
            "codeLens": {
                "codeLensKind": {
                    "valueSet": [
                        "references",
                        "association",
                        "open.uri"
                    ]
                }
            },
            "actionableNotificationSupport": true,
            "openSettingsCommandSupport": true,
            "bindingWizardSupport": true,
            "shouldLanguageServerExitOnShutdown": true
        }
    },
    "trace": "verbose",
    "workspaceFolders": [
        {
            "uri": "file:///c%3A/Users/MyUser/MyProject",
            "name": "MyProject"
        }
    ]
}


Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.
Unrecognized VM option 'ExitOnOutOfMemoryError'
Did you mean 'OnOutOfMemoryError=<value>'?
[Info  - 15:36:24] Connection to server got closed. Server will restart.
[Trace - 15:36:24] Sending request 'initialize - (0)'.
```
This pattern seemingly repeats 5 times total, and finally exits with:
```
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.
Unrecognized VM option 'ExitOnOutOfMemoryError'
Did you mean 'OnOutOfMemoryError=<value>'?
[Error - 15:36:24] Connection to server got closed. Server will not be restarted.
```

**Comment by angelozerr** [comment]
@Torsteinws thanks for the trace.

I suspect the problem comes from the use of Java `C:\\Applications\\Atlassian\\atlassian-plugin-sdk-8.0.16\\jre`

@rgrunber have you some idea?

@Torsteinws I suggest that you add in your settings `"xml.server.preferBinary": true`

**Comment by fbricon** [comment]
looks like C:\Applications\Atlassian\atlassian-plugin-sdk-8.0.16\jre doesn't support the  ExitOnOutOfMemoryError flag, whether the Java version is too old or the JRE generally doesn't support it (which I doubt).

As @angelozerr said, "xml.server.preferBinary": true is a possible solution, alternatively, you can point the "xml.java.home" preference to a more recent version of a JRE/JDK.

**Comment by Torsteinws** [comment]
Adding `"xml.server.preferBinary": true` to settings solved my problems! 

Everything seems to work now. Thank you so much!

**Comment by fbricon** [comment]
@rgrunber @angelozerr  If the JVM server failed, we should prolly automatically switch to the Binary one, wdyt?

**Comment by rgrunber** [comment]
Instead of `CloseAction.DoNotRestart` when there's many failed restarts, maybe we could try to set `preferBinary` and issue a `CloseAction.Restart`

**Comment by Torsteinws** [comment]
I would just like to say from an end user perspective that I found it very difficult to troubleshoot this problem. I would have probably just given up on the extension if I hadn't found help here.

So I think it would be really awesome if you addressed this in some sort of way!  :-)

**Comment by rgrunber** [comment]
I have filed https://github.com/redhat-developer/vscode-xml/issues/652 .

Maybe we could include a link to the troubleshooting docs in any fatal errors if you don't already ?

**Comment by angelozerr** [comment]
> I would just like to say from an end user perspective that I found it very difficult to troubleshoot this problem. I would have probably just given up on the extension if I hadn't found help here.

> So I think it would be really awesome if you addressed this in some sort of way! :-)

@Torsteinws I'm glad you fixed your issue with binary. Your feedback is really valuable for us in order to vscode-xml user are happy with it.

> I have filed https://github.com/redhat-developer/vscode-xml/issues/652 .

Thanks @rgrunber !

**Comment by yinwenhao** [comment]
I encountered the same problem, and solved it in the same way. I guess it's related to the JDK version on my machine. I noticed the XML extension is using Java 17 JVM, while I have JDK 18 installed locally. The output now:
[Info  - 18:09:37] Jul 25, 2023 06:09:37 org.eclipse.lemminx.XMLLanguageServer initialize()
Message: Initializing XML Language server
LemMinX Server info:
 - Version : 0.26.1
 - Native Image
 - VM Version : 17.0.2
 - Git 8379294 - [maven-release-plugin] prepare release 0.26.1
[Warn  - 18:09:38] Jul 25, 2023 06:09:38 org.eclipse.lsp4j.jsonrpc.RemoteEndpoint handleCancellation()
Message: Unmatched cancel notification for request id 6
[Warn  - 18:09:38] Jul 25, 2023 06:09:38 org.eclipse.lsp4j.jsonrpc.RemoteEndpoint handleCancellation()
Message: Unmatched cancel notification for request id 5
